**Windows** [Build 5233](https://ci.wildfly.org/viewLog.html?buildId=24090&buildTypeId=WF_PullRequest_Windows) outcome was **FAILURE** using a merge of b30367ac879687c7129bf525a09597a5b190e072
Summary: Tests failed: 1 (1 new), passed: 3421, ignored: 195 Build time: 01:39:01
### Failed tests
```
org.jboss.as.test.clustering.cluster.web.ConcurrentFineWebFailoverTestCase(SYNC-tcp).testGracefulUndeployFailover: java.lang.AssertionError: Session failed to replicate after container 1 was shutdown. expected:<3> but was:<2>
	at org.jboss.as.test.clustering.cluster.web.AbstractWebFailoverTestCase.testFailover(AbstractWebFailoverTestCase.java:155)
	at org.jboss.as.test.clustering.cluster.web.AbstractWebFailoverTestCase.testGracefulUndeployFailover(AbstractWebFailoverTestCase.java:108)
------- Stdout: -------
18:20:31,816 INFO  [org.jboss.as.test.clustering.NodeUtil] (main) Starting deployment=NONE, container=container-0
18:20:31,817 INFO  [org.jboss.as.test.clustering.NodeUtil] (main) Starting deployment=NONE, container=container-1
18:20:31,817 INFO  [org.jboss.as.test.clustering.NodeUtil] (main) Deploying deployment=deployment-0
18:20:31,855 INFO  [org.jboss.as.repository] (management-handler-thread - 1) WFLYDR0001: Content added at location C:\BuildAgent\work\a31d203e70e89f90\testsuite\integration\clustering\target\wildfly-SYNC-tcp-0\standalone\data\content\2e\b0c5f06f4f87cab865df04fe94d090fc3ec0ea\content
18:20:31,861 INFO  [org.jboss.as.server.deployment] (MSC service thread 1-8) WFLYSRV0027: Starting deployment of "fine-concurrent-distributable.war" (runtime-name: "fine-concurrent-distributable.war")
18:20:31,926 WARN  [org.jboss.as.dependency.private] (MSC service thread 1-6) WFLYSRV0018: Deployment "deployment.fine-concurrent-distributable.war" is using a private module ("org.jboss.as.clustering.common:main") which may be changed or removed in future versions without notice.
18:20:31,926 WARN  [org.jboss.as.dependency.private] (MSC service thread 1-6) WFLYSRV0018: Deployment "deployment.fine-concurrent-distributable.war" is using a private module ("org.jboss.as.server:main") which may be changed or removed in future versions without notice.
18:20:31,927 WARN  [org.jboss.as.dependency.private] (MSC service thread 1-6) WFLYSRV0018: Deployment "deployment.fine-concurrent-distributable.war" is using a private module ("org.infinispan:main") which may be changed or removed in future versions without notice.
18:20:31,938 INFO  [org.jboss.weld.deployer] (MSC service thread 1-6) WFLYWELD0003: Processing weld deployment fine-concurrent-distributable.war
18:20:31,957 INFO  [org.jboss.as.ejb3.deployment] (MSC service thread 1-6) WFLYEJB0473: JNDI bindings for session bean named 'TopologyChangeListenerBean' in deployment unit 'deployment "fine-concurrent-distributable.war"' are as follows:

	java:global/fine-concurrent-distributable/TopologyChangeListenerBean!org.jboss.as.test.clustering.TopologyChangeListener
	java:app/fine-concurrent-distributable/TopologyChangeListenerBean!org.jboss.as.test.clustering.TopologyChangeListener
	java:module/TopologyChangeListenerBean!org.jboss.as.test.clustering.TopologyChangeListener
	java:jboss/exported/fine-concurrent-distributable/TopologyChangeListenerBean!org.jboss.as.test.clustering.TopologyChangeListener
	java:global/fine-concurrent-distributable/TopologyChangeListenerBean
	java:app/fine-concurrent-distributable/TopologyChangeListenerBean
	java:module/TopologyChangeListenerBean

18:20:32,015 INFO  [org.infinispan.configuration.cache.EvictionConfigurationBuilder] (ServerService Thread Pool -- 14) ISPN000152: Passivation configured without an eviction policy being selected. Only manually evicted entities will be passivated.
18:20:32,016 INFO  [org.infinispan.configuration.cache.EvictionConfigurationBuilder] (ServerService Thread Pool -- 14) ISPN000152: Passivation configured without an eviction policy being selected. Only manually evicted entities will be passivated.
18:20:32,031 INFO  [org.infinispan.configuration.cache.EvictionConfigurationBuilder] (ServerService Thread Pool -- 32) ISPN000152: Passivation configured without an eviction policy being selected. Only manually evicted entities will be passivated.
18:20:32,032 INFO  [org.infinispan.configuration.cache.EvictionConfigurationBuilder] (ServerService Thread Pool -- 32) ISPN000152: Passivation configured without an eviction policy being selected. Only manually evicted entities will be passivated.
18:20:32,041 INFO  [org.infinispan.configuration.cache.EvictionConfigurationBuilder] (ServerService Thread Pool -- 73) ISPN000152: Passivation configured without an eviction policy being selected. Only manually evicted entities will be passivated.
18:20:32,042 INFO  [org.infinispan.configuration.cache.EvictionConfigurationBuilder] (ServerService Thread Pool -- 73) ISPN000152: Passivation configured without an eviction policy being selected. Only manually evicted entities will be passivated.
18:20:32,045 INFO  [org.jboss.as.clustering.infinispan] (ServerService Thread Pool -- 14) WFLYCLINF0002: Started client-mappings cache from ejb container
18:20:32,069 INFO  [org.jboss.as.clustering.infinispan] (ServerService Thread Pool -- 32) WFLYCLINF0002: Started routing cache from web container
18:20:32,079 INFO  [org.jboss.as.clustering.infinispan] (ServerService Thread Pool -- 73) WFLYCLINF0002: Started fine-concurrent-distributable.war cache from web container
18:20:32,519 INFO  [org.wildfly.extension.undertow] (ServerService Thread Pool -- 14) WFLYUT0021: Registered web context: /fine-concurrent-distributable
18:20:32,577 INFO  [org.jboss.as.server] (management-handler-thread - 1) WFLYSRV0010: Deployed "fine-concurrent-distributable.war" (runtime-name : "fine-concurrent-distributable.war")
18:20:32,926 INFO  [org.jboss.as.test.clustering.NodeUtil] (main) Deploying deployment=deployment-1
18:20:33,083 INFO  [org.jboss.as.repository] (management-handler-thread - 3) WFLYDR0001: Content added at location C:\BuildAgent\work\a31d203e70e89f90\testsuite\integration\clustering\target\wildfly-SYNC-tcp-1\standalone\data\content\34\3b4e68492a188f65770dd794fe8677db244d02\content
18:20:33,106 INFO  [org.jboss.as.server.deployment] (MSC service thread 1-5) WFLYSRV0027: Starting deployment of "fine-concurrent-distributable.war" (runtime-name: "fine-concurrent-distributable.war")
18:20:33,178 WARN  [org.jboss.as.dependency.private] (MSC service thread 1-2) WFLYSRV0018: Deployment "deployment.fine-concurrent-distributable.war" is using a private module ("org.jboss.as.clustering.common:main") which may be changed or removed in future versions without notice.
18:20:33,179 WARN  [org.jboss.as.dependency.private] (MSC service thread 1-2) WFLYSRV0018: Deployment "deployment.fine-concurrent-distributable.war" is using a private module ("org.jboss.as.server:main") which may be changed or removed in future versions without notice.
18:20:33,179 WARN  [org.jboss.as.dependency.private] (MSC service thread 1-2) WFLYSRV0018: Deployment "deployment.fine-concurrent-distributable.war" is using a private module ("org.infinispan:main") which may be changed or removed in future versions without notice.
18:20:33,185 INFO  [org.jboss.weld.deployer] (MSC service thread 1-2) WFLYWELD0003: Processing weld deployment fine-concurrent-distributable.war
18:20:33,198 INFO  [org.jboss.as.ejb3.deployment] (MSC service thread 1-2) WFLYEJB0473: JNDI bindings for session bean named 'TopologyChangeListenerBean' in deployment unit 'deployment "fine-concurrent-distributable.war"' are as follows:

	java:global/fine-concurrent-distributable/TopologyChangeListenerBean!org.jboss.as.test.clustering.TopologyChangeListener
	java:app/fine-concurrent-distributable/TopologyChangeListenerBean!org.jboss.as.test.clustering.TopologyChangeListener
	java:module/TopologyChangeListenerBean!org.jboss.as.test.clustering.TopologyChangeListener
	java:jboss/exported/fine-concurrent-distributable/TopologyChangeListenerBean!org.jboss.as.test.clustering.TopologyChangeListener
	java:global/fine-concurrent-distributable/TopologyChangeListenerBean
	java:app/fine-concurrent-distributable/TopologyChangeListenerBean
	java:module/TopologyChangeListenerBean



```

--- a/pom.xml
+++ b/pom.xml
@@ -218,7 +218,7 @@
         <version.org.wildfly.build-tools>1.1.3.Final</version.org.wildfly.build-tools>
         <version.org.wildfly.checkstyle-config>1.0.5.Final</version.org.wildfly.checkstyle-config>
         <version.org.wildfly.core>3.0.0.Alpha5</version.org.wildfly.core>
-        <version.org.wildfly.arquillian>1.0.2.Final</version.org.wildfly.arquillian>
+        <version.org.wildfly.arquillian>2.0.0.Final</version.org.wildfly.arquillian>
         <version.org.yaml.snakeyaml>1.15</version.org.yaml.snakeyaml>
         <version.sun.jaxb>2.2.11.jbossorg-1</version.sun.jaxb>
         <version.sun.saaj-impl>1.3.16-jbossorg-1</version.sun.saaj-impl>
@@ -6024,17 +6024,11 @@
                 <type>pom</type>
                 <scope>import</scope>
             </dependency>
+
             <dependency>
                 <groupId>org.wildfly.arquillian</groupId>
                 <artifactId>wildfly-arquillian-common</artifactId>
                 <version>${version.org.wildfly.arquillian}</version>
-            </dependency>
-
-            <!-- only needed until wildfly-arquillian has properly synced this version with arquillian it uses -->
-            <dependency>
-                <groupId>org.jboss.shrinkwrap.descriptors</groupId>
-                <artifactId>shrinkwrap-descriptors-impl-base</artifactId>
-                <version>${version.org.jboss.shrinkwrap.descriptors}</version>
             </dependency>
 
             <dependency>

--- a/testsuite/integration/clustering/src/test/java/org/jboss/as/test/clustering/ClusteringTestConstants.java
+++ b/testsuite/integration/clustering/src/test/java/org/jboss/as/test/clustering/ClusteringTestConstants.java
@@ -76,6 +76,7 @@ public interface ClusteringTestConstants {
      * Timeouts.
      */
     int GRACE_TIME_TO_REPLICATE = TimeoutUtil.adjust(3000);
+    int GRACEFUL_SHUTDOWN_TIMEOUT = TimeoutUtil.adjust(15);
 
     /**
      * TODO: This will be removed.

--- a/testsuite/integration/clustering/src/test/java/org/jboss/as/test/clustering/NodeUtil.java
+++ b/testsuite/integration/clustering/src/test/java/org/jboss/as/test/clustering/NodeUtil.java
@@ -23,6 +23,7 @@ package org.jboss.as.test.clustering;
 
 import org.jboss.arquillian.container.test.api.ContainerController;
 import org.jboss.arquillian.container.test.api.Deployer;
+import org.jboss.as.arquillian.api.WildFlyContainerController;
 import org.jboss.logging.Logger;
 
 /**
@@ -78,6 +79,18 @@ public final class NodeUtil {
                 log.info("Stopping container=" + container);
                 controller.stop(container);
                 log.info("Stopped container=" + container);
+            } catch (Throwable e) {
+                log.error("Failed to stop containers", e);
+            }
+        }
+    }
+
+    public static void stop(int timeout, WildFlyContainerController controller, String... containers) {
+        for (String container : containers) {
+            try {
+                log.info("Stopping container=" + container + ", timeout=" + timeout);
+                controller.stop(container, timeout);
+                log.info("Stopped container=" + container + ", timeout=" + timeout);
             } catch (Throwable e) {
                 log.error("Failed to stop containers", e);
             }

--- a/testsuite/integration/clustering/src/test/java/org/jboss/as/test/clustering/cluster/ClusterAbstractTestCase.java
+++ b/testsuite/integration/clustering/src/test/java/org/jboss/as/test/clustering/cluster/ClusterAbstractTestCase.java
@@ -26,10 +26,10 @@ import java.util.TreeMap;
 
 import org.apache.http.Header;
 import org.apache.http.HttpResponse;
-import org.jboss.arquillian.container.test.api.ContainerController;
 import org.jboss.arquillian.container.test.api.Deployer;
 import org.jboss.arquillian.container.test.api.RunAsClient;
 import org.jboss.arquillian.test.api.ArquillianResource;
+import org.jboss.as.arquillian.api.WildFlyContainerController;
 import org.jboss.as.test.clustering.ClusteringTestConstants;
 import org.jboss.as.test.clustering.NodeUtil;
 import org.jboss.as.web.session.RoutingSupport;
@@ -75,7 +75,7 @@ public abstract class ClusterAbstractTestCase implements ClusteringTestConstants
     }
 
     @ArquillianResource
-    protected ContainerController controller;
+    protected WildFlyContainerController controller;
     @ArquillianResource
     protected Deployer deployer;
 
@@ -109,6 +109,10 @@ public abstract class ClusterAbstractTestCase implements ClusteringTestConstants
 
     protected void stop(String... containers) {
         NodeUtil.stop(this.controller, containers);
+    }
+
+    protected void stop(int timeout, String... containers) {
+        NodeUtil.stop(timeout, this.controller, containers);
     }
 
     protected void deploy(String... deployments) {
@@ -153,7 +157,7 @@ public abstract class ClusterAbstractTestCase implements ClusteringTestConstants
             ClusterAbstractTestCase.this.stop(this.getContainers(nodes));
         }
 
-        private String[] getContainers(String... nodes) {
+        String[] getContainers(String... nodes) {
             String[] containers = new String[nodes.length];
             for (int i = 0; i < nodes.length; ++i) {
                 String node = nodes[i];
@@ -164,6 +168,13 @@ public abstract class ClusterAbstractTestCase implements ClusteringTestConstants
                 containers[i] = container;
             }
             return containers;
+        }
+    }
+
+    public class GracefulRestartLifecycle extends RestartLifecycle {
+        @Override
+        public void stop(String... nodes) {
+            ClusterAbstractTestCase.this.stop(GRACEFUL_SHUTDOWN_TIMEOUT, this.getContainers(nodes));
         }
     }
 

--- a/testsuite/integration/clustering/src/test/java/org/jboss/as/test/clustering/cluster/ejb/remote/RemoteFailoverTestCase.java
+++ b/testsuite/integration/clustering/src/test/java/org/jboss/as/test/clustering/cluster/ejb/remote/RemoteFailoverTestCase.java
@@ -290,10 +290,18 @@ public class RemoteFailoverTestCase extends ClusterAbstractTestCase {
         }
     }
 
-    @InSequence(3)
     @Test
-    @Ignore("WFLY-3532")
-    public void testConcurrentFailover() throws Exception {
+    public void testGracefulShutdownConcurrentFailover() throws Exception {
+        this.testConcurrentFailover(new GracefulRestartLifecycle());
+    }
+
+    @Test
+    @Ignore("Needs graceful undeploy support")
+    public void testGracefulUndeployConcurrentFailover() throws Exception {
+        this.testConcurrentFailover(new RedeployLifecycle());
+    }
+
+    public void testConcurrentFailover(Lifecycle lifecycle) throws Exception {
         ContextSelector<EJBClientContext> selector = EJBClientContextSelector.setup(CLIENT_PROPERTIES);
         try (EJBDirectory directory = new RemoteEJBDirectory(MODULE_NAME)) {
             Incrementor bean = directory.lookupStateful(SlowToDestroyStatefulIncrementorBean.class, Incrementor.class);
@@ -310,7 +318,7 @@ public class RemoteFailoverTestCase extends ClusterAbstractTestCase {
                 Future<?> future = executor.scheduleWithFixedDelay(new IncrementTask(bean, count, latch), 0, INVOCATION_WAIT, TimeUnit.MILLISECONDS);
                 latch.await();
 
-                undeploy(this.findDeployment(target));
+                lifecycle.stop(target);
 
                 future.cancel(false);
                 try {
@@ -319,28 +327,13 @@ public class RemoteFailoverTestCase extends ClusterAbstractTestCase {
                     // Ignore
                 }
 
-                deploy(this.findDeployment(target));
-
-                latch = new CountDownLatch(1);
-                future = executor.scheduleWithFixedDelay(new IncrementTask(bean, count, latch), 0, INVOCATION_WAIT, TimeUnit.MILLISECONDS);
-                latch.await();
-
-                stop(this.findContainer(target));
-
-                future.cancel(false);
-                try {
-                    future.get();
-                } catch (CancellationException e) {
-                    // Ignore
-                }
-
-                start(this.findContainer(target));
+                lifecycle.start(target);
 
                 latch = new CountDownLatch(1);
                 future = executor.scheduleWithFixedDelay(new LookupTask(directory, SlowToDestroyStatefulIncrementorBean.class, latch), 0, INVOCATION_WAIT, TimeUnit.MILLISECONDS);
                 latch.await();
 
-                undeploy(this.findDeployment(target));
+                lifecycle.stop(target);
 
                 future.cancel(false);
                 try {
@@ -349,22 +342,7 @@ public class RemoteFailoverTestCase extends ClusterAbstractTestCase {
                     // Ignore
                 }
 
-                deploy(this.findDeployment(target));
-
-                latch = new CountDownLatch(1);
-                future = executor.scheduleWithFixedDelay(new LookupTask(directory, SlowToDestroyStatefulIncrementorBean.class, latch), 0, INVOCATION_WAIT, TimeUnit.MILLISECONDS);
-                latch.await();
-
-                stop(this.findContainer(target));
-
-                future.cancel(false);
-                try {
-                    future.get();
-                } catch (CancellationException e) {
-                    // Ignore
-                }
-
-                start(this.findContainer(target));
+                lifecycle.start(target);
             } finally {
                 executor.shutdownNow();
             }

